### PR TITLE
strutil/shlex: add Quote() and Join() for shell-escaped quoting

### PR DESCRIPTION
**Pull request by rebornplusplus**
Inspired from Python shlex [0, 1], add Quote(string) to return a shell-escaped version of a string.
Add Join([]string) to join the shell-escaped version of the strings in slice.

References:
- [0] https://github.com/python/cpython/blob/fd155b91392c5bd238781bb11e040cf46c94725c/Lib/shlex.py#L323C31-L334
- [1] https://github.com/python/cpython/blob/fd155b91392c5bd238781bb11e040cf46c94725c/Lib/test/test_shlex.py#L327-L351

### CODE REVIEW

File: strutil/shlex/shlex.go [hunk 1]
```diff
@@ -414,3 +415,28 @@ func Split(s string) ([]string, error) {
...
+// Join returns a shell-escaped string version of the slice s
+func Join(s []string) string {
+	var quoted []string
+	for _, subString := range s {
+		quoted = append(quoted, Quote(subString))
+	}
+	return strings.Join(quoted, " ")
+}
+
+var unsafeQuoteChars = regexp.MustCompile(`[^\w@%+=:,./-]`)
```
Comment (nvinuesa) [inline strutil/shlex/shlex.go:428]: maybe you could transform `unsafeQuoteChars` to a `const`?

**Review by nvinuesa — APPROVED** [review]
Apart from a small comment on the var -> const, LGTM

File: strutil/shlex/shlex.go (hunk 1 above)
Comment (troyanov) [reply strutil/shlex/shlex.go:428]: > maybe you could transform `unsafeQuoteChars` to a `const`?

IIRC `regexp.MustCompile` will return `*regexp.Regexp` which cannot be a constant

File: strutil/shlex/shlex.go [hunk 2]
```diff
@@ -414,3 +415,28 @@ func Split(s string) ([]string, error) {
 		subStrings = append(subStrings, word)
 	}
 }
+
+// Join returns a shell-escaped string version of the slice s
+func Join(s []string) string {
+	var quoted []string
```
Comment (troyanov) [inline strutil/shlex/shlex.go:421]: I think you can create a slice of a known size here, to reduce the number of possible allocations during append (when it will grow)
```go
quoted := make([]string, len(s))
```

And then in the `for-loop` just set the value of `quoted[i]`

Comment (troyanov) [inline strutil/shlex/shlex.go:421]: I think you can create a slice of a known size here, to reduce the number of possible allocations during append (when it will grow)
```go
quoted := make([]string, len(s))
```

And then in the `for-loop` just set the value of `quoted[i]`

**Review by troyanov — CHANGES_REQUESTED** [review]

File: strutil/shlex/shlex.go (hunk 2 above)
Comment (rebornplusplus) [reply strutil/shlex/shlex.go]: Thank you. Done.

File: strutil/shlex/shlex.go [hunk 3]
```diff
@@ -414,3 +415,28 @@ func Split(s string) ([]string, error) {
 		subStrings = append(subStrings, word)
 	}
 }
+
+// Join returns a shell-escaped string version of the slice s
+func Join(s []string) string {
+	quoted := make([]string, len(s))
+	for i := range s {
+		quoted[i] = Quote(s[i])
```
Comment (benhoyt) [inline strutil/shlex/shlex.go:423]: Nit: you can use the two-item `range` here to avoid `s[i]`:

```go
for i, x := range s {
    quotes[i] = Quote(x)
}
```

File: strutil/shlex/shlex.go [hunk 4]
```diff
@@ -414,3 +415,28 @@ func Split(s string) ([]string, error) {
 		subStrings = append(subStrings, word)
 	}
 }
+
+// Join returns a shell-escaped string version of the slice s
```
Comment (benhoyt) [inline strutil/shlex/shlex.go:419]:
```suggestion
// Join returns a shell-escaped string version of the arguments in s.
```

File: strutil/shlex/shlex.go [hunk 5]
```diff
@@ -414,3 +415,28 @@ func Split(s string) ([]string, error) {
...
+	quoted := make([]string, len(s))
+	for i := range s {
+		quoted[i] = Quote(s[i])
+	}
+	return strings.Join(quoted, " ")
+}
+
+var unsafeQuoteChars = regexp.MustCompile(`[^\w@%+=:,./-]`)
+
+// Quote returns a shell-escaped version of the string s
```
Comment (benhoyt) [inline strutil/shlex/shlex.go:430]:
```suggestion
// Quote returns a shell-escaped version of the single argument s.
```

File: strutil/shlex/shlex.go [hunk 6]
```diff
@@ -414,3 +415,28 @@ func Split(s string) ([]string, error) {
...
+func Quote(s string) string {
+	if s == "" {
+		return `''`
+	}
+	if unsafeQuoteChars.Find([]byte(s)) == nil {
```
Comment (benhoyt) [inline strutil/shlex/shlex.go:435]: Probably better to use `.FindString(s)` and compare `== ""` to avoid the allocation/conversion to `[]byte`. (The `""` case has been checked above so doesn't matter here.)

File: strutil/shlex/shlex_test.go [hunk 7]
```diff
@@ -157,3 +157,69 @@ func TestNastyReader(t *testing.T) {
...
+		expected: `''`,
+	}, {
+		input:    safeUnquoted,
+		expected: safeUnquoted,
+	}, {
+		input:    `foo bar xyz`,
+		expected: `'foo bar xyz'`,
+	}}
+
+	for _, u := range unsafe {
```
Comment (benhoyt) [inline strutil/shlex/shlex_test.go:182]: Given we're doing table-driven tests anyway, I think it's clearer to specify a few cases with literals inline rather than having this loop which appends to `quoteTests` -- it's a bit hard to understand what's actually being tested as is.

File: strutil/shlex/shlex_test.go [hunk 8]
```diff
@@ -157,3 +157,69 @@ func TestNastyReader(t *testing.T) {
...
+		quoteTests = append(quoteTests, testInfo{
+			input:    `foo` + string(u) + `'bar'`,
+			expected: `'foo` + string(u) + `'"'"'bar'"'"''`,
+		})
+	}
+
+	for _, test := range quoteTests {
+		quoted := Quote(test.input)
+		if quoted != test.expected {
+			t.Errorf("expected %s, got %s", test.expected, quoted)
```
Comment (benhoyt) [inline strutil/shlex/shlex_test.go:196]: With table-driven tests it's best to have some way of knowing which test that failed. It's usually best to use `t.Run()` inside the loop and explicitly give the test a name in your struct, like

```go
quoteTests := []testInfo{{
		name:     "empty",
		input:    ``,
		expected: `''`,
	}, {
	...
```

Or you could avoid the explicit name and make the `t.Run` name `fmt.Sprintf("Quote(%q)", test.input)`.

File: strutil/shlex/shlex_test.go [hunk 9]
```diff
@@ -157,3 +157,69 @@ func TestNastyReader(t *testing.T) {
...
+	}, {
+		input:    []string{`a`, ` b`},
+		expected: `a ' b'`,
+	}, {
+		input:    []string{`a`, ` `, `b`},
+		expected: `a ' ' b`,
+	}, {
+		input:    []string{`"a`, `b"`},
+		expected: `'"a' 'b"'`,
+	}}
```
Comment (benhoyt) [inline strutil/shlex/shlex_test.go:217]: I think we should also test the edge cases of empty slice and nil slice (should be same as empty). Probably also a case where a single argument is like `x y`, and where a single is more than one character like `/foo/bar`. We don't have to be exhaustive as other things should be caught in the `Quote` tests, but just a couple more cases here would be good.

**Review by benhoyt — APPROVED** [review]
Approved but a few minor comments.

File: strutil/shlex/shlex.go (hunk 6 above)
Comment (rebornplusplus) [reply strutil/shlex/shlex.go]: Thanks. Done.

File: strutil/shlex/shlex_test.go (hunk 7 above)
Comment (rebornplusplus) [reply strutil/shlex/shlex_test.go]: Thanks, I have updated the tests.

File: strutil/shlex/shlex_test.go (hunk 9 above)
Comment (rebornplusplus) [reply strutil/shlex/shlex_test.go]: Added a few cases more. Thank you.

**Review by woky — APPROVED** [review]

**Comment by flotter** [comment]
@rebornplusplus please let me know if you struggle with failing tests. Snapd has moved to Go v1.18, so I think we can now remove the Go v1.13 test, and support v1.14 for Pebble and v1.18 for Snapd and Chisel only.

File: strutil/shlex/shlex.go [hunk 10]
```diff
@@ -414,3 +415,28 @@ func Split(s string) ([]string, error) {
...
+func Quote(s string) string {
+	if s == "" {
+		return `''`
+	}
+	if unsafeQuoteChars.FindString(s) == "" {
+		return s
+	}
+
+	// use single quotes, and put single quotes into double quotes
+	// the string $'b is then quoted as '$'"'"'b'
```
[the remark this answers is not in the document]
Comment (rebornplusplus) [reply strutil/shlex/shlex.go]: Understood. Thanks, dropped the comments.

File: strutil/shlex/shlex.go [hunk 11]
```diff
@@ -414,3 +415,28 @@ func Split(s string) ([]string, error) {
...
+}
+
+var unsafeQuoteChars = regexp.MustCompile(`[^\w@%+=:,./-]`)
+
+// Quote returns a shell-escaped version of the single argument s.
+func Quote(s string) string {
+	if s == "" {
+		return `''`
+	}
+	if unsafeQuoteChars.FindString(s) == "" {
```
[the remark this answers is not in the document]
Comment (rebornplusplus) [reply strutil/shlex/shlex.go]: Thanks, updated in https://github.com/canonical/x-go/commit/137eca1c0c0068c2bf11bee2aa7876041c835057.

File: strutil/shlex/shlex.go [hunk 12]
```diff
@@ -414,3 +414,30 @@ func Split(s string) ([]string, error) {
...
+		return `''`
+	}
+	for _, c := range s {
+		switch {
+		case c >= 'a' && c <= 'z':
+		case c >= 'A' && c <= 'Z':
+		case c >= '0' && c <= '9':
+		case strings.ContainsRune("_@%+=:,./-", c):
+		default:
+			return "'" + strings.ReplaceAll(s, "'", "'\"'\"'") + "'"
```
Comment (benhoyt) [inline strutil/shlex/shlex.go:439]: For whatever reason it took me a while to figure out that these were the safe characters. Maybe add a small comment like so?

```go
default:
    // The above cases are all the safe characters. If there are any other characters,
    // quote the entire input.
    return "'" + strings.ReplaceAll(s, "'", "'\"'\"'") + "'"
```

File: strutil/shlex/shlex.go [hunk 13]
```diff
@@ -414,3 +414,30 @@ func Split(s string) ([]string, error) {
...
+	}
+	return strings.Join(quoted, " ")
+}
+
+// Quote returns a shell-escaped version of the single argument s.
+func Quote(s string) string {
+	if s == "" {
+		return `''`
+	}
+	for _, c := range s {
```
Comment (benhoyt) [inline strutil/shlex/shlex.go:432]: This works fine, but also decodes the UTF-8 input into runes, and there's no reason to pay that penalty here. I suggest processing it as bytes instead. I think this would do it:

```
for i := 0; i < len(s); i++ {
    c := s[i]
    switch {
		case c >= 'a' && c <= 'z':
		case c >= 'A' && c <= 'Z':
		case c >= '0' && c <= '9':
		case strings.IndexByte("_@%+=:,./-", c) >= 0:
		...
    }
}

**Review by benhoyt — APPROVED** [review]
Approved but with a couple of minor mods suggested.

File: strutil/shlex/shlex.go (hunk 12 above)
Comment (rebornplusplus) [reply strutil/shlex/shlex.go:442]: Thanks for the wording Ben. Updated in https://github.com/canonical/x-go/pull/19/commits/08c9987a9a6f53281d401846358666dcb4c07a12.

File: strutil/shlex/shlex.go (hunk 13 above)
Comment (rebornplusplus) [reply strutil/shlex/shlex.go]: Thanks for the insight! Update in https://github.com/canonical/x-go/pull/19/commits/08c9987a9a6f53281d401846358666dcb4c07a12.

Comment (niemeyer) [reply strutil/shlex/shlex.go:432]: Looks okay with Ben's comments, but why are those not part of the PR still?

Comment (rebornplusplus) [reply strutil/shlex/shlex.go]: Sorry about that @niemeyer.
I updated them in https://github.com/canonical/x-go/commit/08c9987a9a6f53281d401846358666dcb4c07a12, but I must have messed something up while rebasing. I should have checked carefully afterwards. I rebased and added the changes in https://github.com/canonical/x-go/pull/19/commits/58568dfafa951c44b6d1b05a01e9d2ab93e0c853.

**Review by niemeyer — APPROVED** [review]
Thanks!
